File: client/src/events/collisions.js
/**
 * 屏幕空间碰撞推挤（纯函数，供 EventBubbles.resolve 与单测复用）。
 *
 * 规则与 EventBubbles 保持一致：
 * - 按年份排序，早出现者优先不动，晚出现者被推挤
 * - fixed 节点（如地图政权标签）不可被推挤，只作为固定障碍
 * - 优先向下推（视觉自然）；垂直将超限时改水平推挤
 *
 * @param {Array<{year:number, rect:{x:number,y:number,w:number,h:number}, fixed?:boolean}>} nodes
 * @param {object} [opts]
 * @param {number} [opts.gap=6]     泡泡间留白（px）
 * @param {number} [opts.maxPush=64] 单方向最大推挤量（px）
 * @returns {Array<{dx:number, dy:number}>} 每个节点的推挤量（与入参顺序一致）
 */
export function resolveCollisions(nodes, { gap = 6, maxPush = 64 } = {}) {
  const result = nodes.map(() => ({ dx: 0, dy: 0 }));
  // 年份升序；固定障碍（year = -Infinity）自然排最前
  const order = nodes
    .map((nd, i) => ({ nd, i }))
    .sort((a, b) => a.nd.year - b.nd.year);

  for (let oi = 0; oi < order.length; oi++) {
    for (let oj = oi + 1; oj < order.length; oj++) {
      const a = order[oi];
      const b = order[oj];
      if (b.nd.fixed) continue; // 障碍物不可被推挤
      const ax = a.nd.rect.x + result[a.i].dx;
      const ay = a.nd.rect.y + result[a.i].dy;
      const bx = b.nd.rect.x + result[b.i].dx;
      const by = b.nd.rect.y + result[b.i].dy;
      const ox = Math.min(ax + a.nd.rect.w, bx + b.nd.rect.w) - Math.max(ax, bx);
      const oy = Math.min(ay + a.nd.rect.h, by + b.nd.rect.h) - Math.max(ay, by);
      if (ox <= 0 || oy <= 0) continue;

      const verticalRoom = maxPush - Math.abs(result[b.i].dy);
      if (oy + gap <= verticalRoom) {
        result[b.i].dy += oy + gap;
      } else {
        const dir = result[b.i].dx <= 0 ? 1 : -1; // 优先向右，已右偏则向左
        const need = Math.min(ox + gap, maxPush);
        result[b.i].dx += need * dir;
      }
    }
  }
  return result;
}

/** 判断两个矩形是否重叠（含零面积边界） */
export function rectsOverlap(a, b) {
  const ox = Math.min(a.x + a.w, b.x + b.w) - Math.max(a.x, b.x);
  const oy = Math.min(a.y + a.h, b.y + b.h) - Math.max(a.y, b.y);
  return ox > 0 && oy > 0;
}

File: android/app/src/main/java/com/historymap/app/Collisions.kt
package com.historymap.app

/**
 * 屏幕空间碰撞推挤（纯函数，翻译 Web 版 client/src/events/collisions.js）。
 *
 * 规则与 Web 版保持一致：
 * - 按年份排序，早出现者优先不动，晚出现者被推挤
 * - fixed 节点（如地图政权标签）不可被推挤，只作为固定障碍
 * - 优先向下推（视觉自然）；垂直将超限时改水平推挤
 */

/** 屏幕空间矩形（CSS 像素） */
data class RectF2(val x: Float, val y: Float, val w: Float, val h: Float)

/** 推挤量（dx 向右、dy 向下为正） */
data class Shift(val dx: Float, val dy: Float)

/** 碰撞节点：年份越小越优先不动；fixed 为不可移动障碍 */
data class CollisionNode(
    val year: Int,
    val rect: RectF2,
    val fixed: Boolean = false,
)

/**
 * @param nodes 参与推挤的节点
 * @param gap 节点间留白（px）
 * @param maxPush 单方向最大推挤量（px）
 * @return 与入参顺序一致的推挤量列表
 */
fun resolveCollisions(nodes: List<CollisionNode>, gap: Float = 6f, maxPush: Float = 64f): List<Shift> {
    val result = MutableList(nodes.size) { Shift(0f, 0f) }
    // 年份升序；固定障碍（year = Int.MIN_VALUE）自然排最前
    val order = nodes.mapIndexed { i, nd -> nd to i }
        .sortedBy { it.first.year }

    for (oi in order.indices) {
        val (a, ai) = order[oi]
        for (oj in oi + 1 until order.size) {
            val (b, bi) = order[oj]
            if (b.fixed) continue // 障碍物不可被推挤
            val ax = a.rect.x + result[ai].dx
            val ay = a.rect.y + result[ai].dy
            val bx = b.rect.x + result[bi].dx
            val by = b.rect.y + result[bi].dy
            val ox = minOf(ax + a.rect.w, bx + b.rect.w) - maxOf(ax, bx)
            val oy = minOf(ay + a.rect.h, by + b.rect.h) - maxOf(ay, by)
            if (ox <= 0f || oy <= 0f) continue

            val verticalRoom = maxPush - kotlin.math.abs(result[bi].dy)
            if (oy + gap <= verticalRoom) {
                result[bi] = Shift(result[bi].dx, result[bi].dy + oy + gap)
            } else {
                val dir = if (result[bi].dx <= 0f) 1f else -1f // 优先向右，已右偏则向左
                val need = minOf(ox + gap, maxPush)
                result[bi] = Shift(result[bi].dx + need * dir, result[bi].dy)
            }
        }
    }
    return result
}

/**
 * 出屏回收：把已推挤的矩形拉回视口内（纯函数，翻译 collisions.js clampNodeToViewport）。
 * @param minY / maxY 纵向安全区（顶栏底 / 时间轴顶），保证泡泡不进入 UI 铬区。
 */
fun clampToViewport(
    rect: RectF2,
    dx: Float,
    dy: Float,
    viewportW: Float,
    viewportH: Float,
    pad: Float = 6f,
    minY: Float = pad,
    maxY: Float = viewportH - pad,
): Shift {
    if (viewportW <= 0f || viewportH <= 0f) return Shift(dx, dy)
    var nx = dx
    var ny = dy
    val x = rect.x + dx
    val y = rect.y + dy
    if (x < pad) nx += pad - x
    else if (x + rect.w > viewportW - pad) nx += viewportW - pad - (x + rect.w)
    if (y < minY) ny += minY - y
    else if (y + rect.h > maxY) ny += maxY - (y + rect.h)
    return Shift(nx, ny)
}
